classes/module/Module.php

```diff
@@ -3436,30 +3442,70 @@ public function isSymfonyContext()
     }
 
     /**
-     * Access the Symfony Container if we are in Symfony Context.
-     * Note: in this case, we must get a container from SymfonyContainer class.
-     * Note: if not in Symfony context, fallback to legacy Container for FO/BO.
+     * @return bool
+     */
+    public function hasSymfonyContainer(): bool
+    {
+        // Check that a container is available but not generated by PrestaShop
+        $container = $this->getContainer();
+
+        return null !== $container && !($container instanceof LegacyContainerInterface);
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasLegacyContainer(): bool
+    {
+        // Check that a container is available but not generated by PrestaShop
+        $container = $this->getContainer();
+
+        return null !== $container && $container instanceof LegacyContainerInterface;
+    }
+
+    /**
+     * Access a service from the container (found from the controller or the context
+     * depending on cases). It uses ContainerFinder to find the appropriate container.
      *
      * @param string $serviceName
      *
-     * @return object|false if a container is not available, it returns false
+     * @return object|false If a container is not available it returns false
+     *
+     * @throws ServiceCircularReferenceException When a circular reference is detected
+     * @throws ServiceNotFoundException When the service is not defined
+     * @throws \Exception
      */
     public function get($serviceName)
     {
-        if ($this->isSymfonyContext()) {
-            if (null === $this->container) {
-                $this->container = SymfonyContainer::getInstance();
-            }
-
-            return $this->container->get($serviceName);
+        try {
+            $container = $this->getContainer();
+        } catch (ContainerNotFoundException $e) {
+            $container = null;
         }
-
-        if ($this->context->controller instanceof Controller) {
-            return $this->context->controller->get($serviceName);
+        if (null !== $container) {
+            return $container->get($serviceName);
         }
 
         return false;
     }
+
+    /**
+     * Return the container depending on the context:
+     *  - Symfony environment: container generated by the kernel
+     *  - Legacy environment admin: container generated by PrestaShop core for admin controllers
+     *  - Legacy environment front: container generated by PrestaShop core for front controllers
+     *
+     * @return ContainerBuilder|ContainerInterface|null
+     */
+    public function getContainer()
```

Comment:
I don't see  why this should be public

```suggestion
    protected function getContainer()
```

Reply:
Alternatively

```suggestion
    protected function getContainer(): ContainerInterface
```

Reply:
If we remove the `hasSymfonyContainer` `hasLegacyContainer` then this one is useful as public to check the container type manually

Besides it's also useful to get a parameter: `$this->getContainer()->getParameter()` unless we introduce some new shortcut `Module::getParameter`

Reply:
Ok then